src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -97,45 +102,119 @@ const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_unoptimized_lo
+  auto& unoptimized_lqp = get_unoptimized_logical_plan();
+
+  ParameterID next_parameter_id(0);
+
+  if (_translation_info.cacheable) {
+    visit_lqp(unoptimized_lqp, [&values, &next_parameter_id](const auto& node) {
+      if (node) {
+        for (auto& root_expression : node->node_expressions) {
+          visit_expression(root_expression, [&values, &next_parameter_id](auto& expression) {
+            split_expression(expression, values, next_parameter_id);
+            return ExpressionVisitation::VisitArguments;
+          });
+        }
+      }
+      return LQPVisitation::VisitInputs;
+    });
+  }
+
+  return unoptimized_lqp;
+}
+
 const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_optimized_logical_plan() {
   if (_optimized_logical_plan) {
     return _optimized_logical_plan;
   }
 
+  const auto started_preoptimization_cache = std::chrono::high_resolution_clock::now();
+
+  std::vector<std::shared_ptr<AbstractExpression>> values;
+  const auto unoptimized_lqp = get_split_unoptimized_logical_plan(values);
+
   // Handle logical query plan if statement has been cached
   if (lqp_cache) {
-    if (const auto cached_plan = lqp_cache->try_get(_sql_string)) {
-      const auto plan = *cached_plan;
+    if (const auto cached_plan = *(lqp_cache->try_get(unoptimized_lqp))) {
+      const auto plan = cached_plan->lqp;
       DebugAssert(plan, "Optimized logical query plan retrieved from cache is empty.");
       // MVCC-enabled and MVCC-disabled LQPs will evict each other
       if (lqp_is_validated(plan) == (_use_mvcc == UseMvcc::Yes)) {
         // Copy the LQP for reuse as the LQPTranslator might modify mutable fields (e.g., cached column_expressions)
         // and concurrent translations might conflict.
-        _optimized_logical_plan = plan->deep_copy();
+        _optimized_logical_plan = cached_plan->instantiate(values);
+        auto olqp = _optimized_logical_plan->deep_copy();
+        _metrics->query_plan_cache_hit = true;
+        _optimized_logical_plan = _pruning_optimizer->optimize(std::move(olqp));
+        const auto done_cache = std::chrono::high_resolution_clock::now();
+        _metrics->cache_duration =
+            std::chrono::duration_cast<std::chrono::nanoseconds>(done_cache - started_preoptimization_cache);
         return _optimized_logical_plan;
       }
     }
   }
 
-  auto unoptimized_lqp = get_unoptimized_logical_plan();
-
-  const auto started = std::chrono::high_resolution_clock::now();
-
   // The optimizer works on the original unoptimized LQP nodes. After optimizing, the unoptimized version is also
   // optimized, which could lead to subtle bugs. optimized_logical_plan holds the original values now.
   // As the unoptimized LQP is only used for visualization, we can afford to recreate it if necessary.
   _unoptimized_logical_plan = nullptr;
 
-  _optimized_logical_plan = _optimizer->optimize(std::move(unoptimized_lqp));
+  auto ulqp = unoptimized_lqp->deep_copy();
```

Comment:
Why do we need the `deep_copy` here? Also, naming.

Reply:
Otherwise, we only get the first node.
We need the complete lqp again in order to move it into the optimizer, where it can be changed without side effects.